src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -312,6 +313,67 @@ impl ScalarFunc {
             _ => Err(box_err!("invalid direction value: {}", direction)),
         }
     }
+
+    #[inline]
+    pub fn to_base64<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string(ctx, row));
+        if encoded_size(s.len()).is_none() {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+        // TODO:
+        // 1. check with maxAllowedPacket
+        // 2. b.tp.Flen == -1 || b.tp.Flen > mysql.MaxBlobWidth
+        let r = base64::encode_config(&s, base64_config());
+        Ok(Some(Cow::Owned(r.into_bytes())))
+    }
+
+    #[cfg_attr(feature = "cargo-clippy", allow(wrong_self_convention))]
+    #[inline]
+    pub fn from_base64<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+
+        let mut input_copy = Vec::<u8>::with_capacity(input.len());
+        input_copy.extend(input.iter().filter(|b| !b" \n\t\r\x0b\x0c".contains(b)));
+        if input_copy.len().checked_mul(3).is_none() || input_copy.len() % 4 != 0 {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+
+        // TODO: check with maxAllowedPacket
+
+        match base64::decode_config(&input_copy, base64_config()) {
+            Ok(r) => Ok(Some(Cow::Owned(r))),
+            _ => Ok(None),
+        }
+    }
+}
+
+#[inline]
+fn base64_config() -> base64::Config {
+    base64::Config::new(
+        base64::CharacterSet::Standard,
+        true,
+        false,
+        base64::LineWrap::Wrap(76, base64::LineEnding::LF),
```

Comment:
It's better to make it `const` and add comments when you need to use such specific number like 76, 3, 4.

Reply:
I'm not quite familiar with rust `const`， I tried before, 
```
const config: base64::Config = base64::Config::new(....)
```
it will tell me `calls in constants are limited to constant functions, tuple structs and tuple variants`
I tried lazy_static, since it will pass to `base64::encode_config`, the type is not matched

Reply:
In rust, you can't initialize a const obj with functions. 
I suggest that you can make these numbers const and add comments to avoid other people feel confused.

Reply:
You can use `let` to create an immutable global variable.

Reply:
@solotzg  means there should be no magic numbers, and it seems you have solved it, good job!